File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-9.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l: int = 0
        r: int = len(nums) - 1

        while l < r:
            if nums[l] < nums[r]:
                return nums[l]
            mid: int = (l + r) // 2

            if nums[mid] > nums[r]:
                # mid > right => right part
                l = mid + 1
            else:
                # else => left part
                r = mid

        return nums[l]

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-9.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // for checking window validity
        Set<Character> unique = new HashSet<>();

        int left = 0;
        int longest = 0;

        for (int right=0; right < s.length(); right++) {
            // while the window is invalid shrink left pointer
            while (unique.contains(s.charAt(right))) {
                unique.remove(s.charAt(left));
                left++;
            }

            unique.add(s.charAt(right));
            longest = Math.max(longest, right-left + 1);
        }
    return longest;
    }

}
